src/ui/src/builder/BuilderFieldsTools.vue

```diff
@@ -0,0 +1,282 @@
+				</button>
+			</div>
+		</div>
+
+		<WdsButton
+			variant="builder"
+			size="small"
+			@click="resetAndShowToolFormModal"
+		>
+			<i class="material-symbols-outlined">add</i>
+			Add tool</WdsButton
+		>
+		<BuilderModal
+			v-if="toolForm.isShown"
+			:close-action="customHandlerModalCloseAction"
+			icon="add"
+			modal-title="Add tool"
+		>
+			<WdsDropdownInput v-model="toolForm.type">
+				<option value="function">Function</option>
+				<option value="graph">Knowledge graph</option>
+			</WdsDropdownInput>
+			<WdsTextInput v-model="toolForm.name"></WdsTextInput>
+			<template v-if="toolForm.type == 'function'">
+				<BuilderEmbeddedCodeEditor
+					v-model="toolForm.code"
+				></BuilderEmbeddedCodeEditor>
+			</template>
+			<template v-if="toolForm.type == 'graph'">
+				<WdsTextInput v-model="toolForm.graphId"></WdsTextInput>
+			</template>
+			<div>
+				<WdsButton variant="builder" @click="saveToolForm"
+					>Save</WdsButton
+				>
+			</div>
+		</BuilderModal>
+	</div>
+</template>
+
+<script setup lang="ts">
+import { toRefs, inject, computed, ref, Ref, ComputedRef } from "vue";
+import { Component, FieldControl } from "@/writerTypes";
+import { useComponentActions } from "./useComponentActions";
+import injectionKeys from "../injectionKeys";
+import WdsButton from "@/wds/WdsButton.vue";
+import BuilderModal, { ModalAction } from "./BuilderModal.vue";
+import WdsTextareaInput from "@/wds/WdsTextareaInput.vue";
+import WdsTextInput from "@/wds/WdsTextInput.vue";
+import WdsDropdownInput from "@/wds/WdsDropdownInput.vue";
+import BuilderEmbeddedCodeEditor from "./BuilderEmbeddedCodeEditor.vue";
+
+type FunctionTool = {
+	type: "function";
+	description: string;
+	parameters: Record<
+		string,
+		{
+			type: string;
+			description: string;
+		}
+	>;
+};
+
+type GraphTool = {
+	type: "graph";
+	graph_ids: string[];
+};
+
+type Tool = FunctionTool | GraphTool;
+
+type ToolForm = {
+	isShown: boolean;
+	type: "function" | "graph";
+	name: string;
+	code: string;
+	graphId: string;
+};
+
+const wf = inject(injectionKeys.core);
+const ssbm = inject(injectionKeys.builderManager);
+const { setContentValue } = useComponentActions(wf, ssbm);
+
+const initFunctionToolCode = `
+{
+	"description": "An example tool",
+	"parameters": {
+		"city": string
+	}
+}
+`.trim();
+
+const toolFormInitValue = {
+	isShown: false,
+	type: "function" as "function" | "graph",
+	name: "new_tool",
+	code: initFunctionToolCode,
+	graphId: "999-999",
+};
+
+const toolForm: Ref<ToolForm> = ref(toolFormInitValue);
+
+const props = defineProps<{
+	componentId: Component["id"];
+	fieldKey: string;
+}>();
+const { componentId, fieldKey } = toRefs(props);
+const component = computed(() => wf.getComponentById(componentId.value));
+
+const tools: ComputedRef<Record<string, Tool>> = computed(() => {
+	let value = {};
+	try {
+		value = JSON.parse(component.value.content[fieldKey.value]);
+	} catch {
+		value = {};
+	}
+	return value;
+});
```

Comment:
Could be a bit shorter
```suggestion
const tools = computed<Record<string, Tool>>(() => {
	try {
		return JSON.parse(component.value.content[fieldKey.value]);
	} catch {
		return {};
	}
});
```